enhancements/ingress/ingress-fault-detection.md

```diff
@@ -0,0 +1,175 @@
+understand why routes on their cluster are not working.
+
+This enhancement also has benefits for internal OpenShift development.
+Often times when networking-related bugs are identified within a cluster, determining which networking sub-component
+(router, SDN, DNS, etc.) is responsible for the observed bug is difficult. Bug reports tend to spend lots of time bouncing between different teams
+as engineers are not sure which component(s) is legitimately at fault. Being able to quickly pinpoint Ingress related issues will expedite the bugfix process.
+
+### Goals
+
+1. Teach the [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) to automatically detect cluster Ingress issues
+and alert cluster administrators when relevant problems are discovered.
+2. Provide a secondary tool that can be triggered at will by a Cluster Administrator or an OpenShift support engineer that spawns an Ingress "debug"
+container that quickly identifies and triages potential Ingress issues within the cluster.
+
+### Non-Goals
+
+1. Teach the Ingress Operator to "self-heal" when cluster Ingress errors or issues are observed.
+2. Modify the [Ingress Controller API](https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go)
+to allow Cluster Administrators to disable the periodic Ingress Operator fault detector at will for a given Ingress Controller.
+
+## Proposal
+
+Once the Ingress Operator has successfully deployed an [Ingress Controller](https://github.com/openshift/cluster-ingress-operator/blob/master/README.md)
+within a cluster, the Ingress Operator will spawn a simple HTTP echo container, such as [Hello Openshift](https://hub.docker.com/r/openshift/hello-openshift/).
+The Ingress Operator will also create the necessary Service and Route resources to provide Ingress to the echo container within the `openshift-ingress-operator`
+namespace over HTTP and HTTP/2 connections. Note that a custom certificate will be needed for any HTTP/2 enabled Routes.
+Then, the Ingress Operator will periodically attempt to send a request to the echo container, and await a proper response. These requests will be sent
+such that they are routed outside of the cluster to the cloud load balancer first (when applicable), in order to ensure that each part of the Ingress "chain" is being tested.
+If the Ingress Operator is unable to send the request to the echo container, the Ingress Operator fails to receive a response
+from the echo container, or the entire request and response flow takes an unusual amount of time, the Ingress Operator will fire an alert via Prometheus.
+Additionally, the echo container will have several endpoint resources associated with it. A static route will be used to infer baseline cluster networking latency
+and jitter. A set of routes that are rotated in and out by periodic creation and deletion will be used to verify that route creation leads to functioning Ingress traffic
```

Comment:
Have we considered to probe those routes via synthetic probes as available in the latest prometheus operator release and therefore coming to OCP sooner than later?

Reply:
No, we have not yet considered synthetic probes. Could you provide a link to some docs or code so I can understand the advantage of using synthetic probes via the Prometheus operator?

Reply:
It was added pretty recently with [this](https://github.com/prometheus-operator/prometheus-operator/pull/2832/files) PR. 
The idea is basically that you can define blackbox probes using https://github.com/prometheus/blackbox_exporter directly via a CR.

This would avoid a need for networking edge to come up with tooling to probe the routes and evaluate those probes.
Instead a synthetic probe would come by default to clusters with the cluster monitoring operator.

cc @s-urbaniak who has more context on prometheus operator and CMO inner workins as well as potential timelines.
Serg, please also correct me if I explained something wrong.

Reply:
@RiRa12621 Thanks for the explanation! For this case, Synthetic Probes sound like a pretty enticing option. Do your or @s-urbaniak have an idea of when we can expect synthetic probes to be added to OCP via the CMO? :smile:

Reply:
@s-urbaniak can probably put a finger on it better, but as far as I can see, it should land in the next promtheus-operator release, which if I'm not wrong should be ready in OCP 4.8, but that's just my external view on it and I would like @s-urbaniak to confirm that.

Reply:
Cool. This enhancement is targeted for 4.7, so we might to start off without synthetic probes, but we could definitely look into switching over to using them in the future :smiley: